cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/AbstractStateChangeProgramController.java

```diff
@@ -0,0 +1,146 @@
+import co.cask.cdap.internal.app.runtime.AbstractProgramController;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.common.util.concurrent.Service;
+import org.apache.twill.common.Threads;
+import org.apache.twill.internal.ServiceListenerAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * Abstract implementation of {@link AbstractProgramController} that responds to state transitions and persists all
+ * state changes.
+ */
+public abstract class AbstractStateChangeProgramController extends AbstractProgramController {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractStateChangeProgramController.class);
+
+  public AbstractStateChangeProgramController(Service service, final ProgramRunId programRunId, final String twillRunId,
+                                              final ProgramStateWriter programStateWriter,
+                                              @Nullable String componentName) {
+    super(programRunId.getParent(), RunIds.fromString(programRunId.getRun()), componentName);
+
+    service.addListener(
+      new ServiceListenerAdapter() {
+        @Override
+        public void starting() {
+          programStateWriter.running(programRunId, twillRunId);
+        }
+
+        @Override
+        public void terminated(Service.State from) {
+          ProgramRunStatus runStatus = ProgramController.State.COMPLETED.getRunStatus();
+          if (from == Service.State.STOPPING) {
+            // Service was killed
+            runStatus = ProgramController.State.KILLED.getRunStatus();
+          }
+          switch (runStatus) {
+            case COMPLETED:
+              programStateWriter.completed(programRunId);
+              break;
+            case KILLED:
+              programStateWriter.killed(programRunId);
+              break;
+            default:
+              // Will never reach this position
+              throw new UnsupportedOperationException(String.format("Cannot terminate program {} for status {}",
+                                                                    programRunId, runStatus));
+          }
+        }
+
+        @Override
+        public void failed(Service.State from, @Nullable final Throwable failure) {
+          programStateWriter.error(programRunId, failure);
+        }
+      },
+      Threads.SAME_THREAD_EXECUTOR
+    );
+  }
+
+  public AbstractStateChangeProgramController(final ProgramRunId programRunId, final String twillRunId,
+                                              final ProgramStateWriter programStateWriter,
+                                              @Nullable String componentName) {
+    super(programRunId.getParent(), RunIds.fromString(programRunId.getRun()), componentName);
+
+    addListener(
+      new AbstractListener() {
+        @Override
+        public void init(ProgramController.State state, @Nullable Throwable cause) {
```

Comment:
@chtyim I feel like this logic is not needed at all anymore. Since the controller has been moved to the runner, the listener won't be added so late as to skip the initial state, right?

Couldn't we remove the `init` method entirely? I was encountering issues with persisting states multiple times and this seems to be to root cause.